File: GFG/DP/44.Largest-Divisible-Subset.cpp
// Largest Divisible Subset

class Solution {
  public:
    vector<int> LargestSubset(int n, vector<int>& arr) {
        // Code here
        sort(arr.begin(),arr.end());
        vector<int> dp(n,1);
        vector<int> hash(n);
        int lastInd = 0;
        for(int idx=0; idx<n; idx++) {
            hash[idx] = idx;
            for(int prev=0; prev<=idx-1; prev++) {
                if(arr[idx]%arr[prev]==0 && 1+dp[prev]>dp[idx]) {
                    dp[idx] = 1 + dp[prev];
                    hash[idx] = prev;
                }
            }
            if(dp[idx] > dp[lastInd]) lastInd = idx;
        }
        
        vector<int> ans;
        ans.push_back(arr[lastInd]);
        while(hash[lastInd] != lastInd) {
            lastInd = hash[lastInd];
            ans.push_back(arr[lastInd]);
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};

File: GFG/DP/44.Largest-Divisible-Subset.java
// Largest Divisible Subset

class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        int n = nums.length;

        Arrays.sort(nums);
        int[] dp = new int[n];
        int[] prev = new int[n];

        for (int i = 0; i < n; i++) {
            dp[i] = 1;
            prev[i] = i;
        }
        
        int lastInd = -1;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if ((1 + dp[j] > dp[i]) && nums[i] % nums[j] == 0) {
                    dp[i] = 1 + dp[j];
                    prev[i] = j;
                }
            }
            if (lastInd == -1 || dp[i] > dp[lastInd]) {
                lastInd = i;
            }
        }

        List<Integer> ans = new ArrayList<>();
        while(prev[lastInd] != lastInd) {
            ans.add(nums[lastInd]);
            lastInd = prev[lastInd];
        }
        ans.add(nums[lastInd]);
        Collections.reverse(ans);

        return ans;
    }
}
